File: Java/heap/더_맵게.java
import java.util.*;

class Solution {
    public int check(Queue<Integer> queue, int K){
        int result = 1;
        for(int i = 0; i < queue.size(); i++){
            int temp = queue.poll();
            if(temp < K){
                result = 0;
                queue.add(temp);
                break;
            }
        }
        return result;
    }
    public int solution(int[] scoville, int K) {   
        int answer = 0;
        Queue<Integer> queue = new PriorityQueue<Integer>();
        
        for(int i = 0; i < scoville.length; i++){
            queue.add(scoville[i]);
        }
        
        try{
            while(check(queue, K) != 1){
                int scovilleScore = 0;
                scovilleScore = queue.poll() + (queue.poll() * 2);
                queue.add(scovilleScore);
                answer++;
            }    
        }catch(NullPointerException e){
            answer = -1;
        }
        return answer;
    }
}

File: C++/heap/더_맵게.cpp
#include <vector>
#include <queue>

using namespace std;

int solution(vector<int> scoville, int K) {
    priority_queue<int, vector<int>, greater<int> > pq;
    int min = 0;
    int secondMin = 0;
    int answer = 0;

    for (int i = 0; i < scoville.size(); i++) {
        pq.push(scoville.at(i));
    }

    if (pq.top() >= K) return -1;

    while (!pq.empty()) {
        answer++;

        if (pq.size() <= 1) {
            answer = 0;
            break;
        }

        min = pq.top();
        pq.pop();
        secondMin = pq.top();
        pq.pop();

        int newScoville = min + (2 * secondMin);

        if (pq.size() == 0 && newScoville >= K) break;
        else if (pq.top() >= K && newScoville >= pq.top()) break;
        else pq.push(newScoville);
    }

    if (answer == 0) answer = -1;

    return answer;
}